node-wasm/src/ports.rs

```diff
@@ -0,0 +1,314 @@
+    request_rx: mpsc::UnboundedReceiver<(ClientId, Result<NodeCommand, TypedMessagePortError>)>,
+}
+
+impl RequestServer {
+    pub fn new() -> RequestServer {
+        let (request_tx, request_rx) = mpsc::unbounded_channel();
+        let (connect_tx, connect_rx) = mpsc::unbounded_channel();
+
+        RequestServer {
+            ports: vec![],
+            connect_tx,
+            connect_rx,
+            request_tx,
+            request_rx,
+        }
+    }
+
+    pub async fn recv(&mut self) -> (ClientId, Result<NodeCommand, TypedMessagePortError>) {
+        loop {
+            select! {
+                message = self.request_rx.recv() => {
+                    return message.expect("request channel should never close");
+                },
+                connection = self.connect_rx.recv() => {
+                    let port = connection.expect("command channel should not close ?");
+                    let client_id = ClientId(self.ports.len());
+                    info!("Connecting client {client_id:?}");
+
+                        match ClientConnection::new(client_id, port, self.request_tx.clone(), self.connect_tx.clone()) {
+                            Ok(port) => self.ports.push(port),
+                            Err(e) => error!("Failed to setup ClientConnection: {e}"),
+                        }
+                }
+            }
+        }
+    }
+
+    pub fn get_connect_channel(&self) -> mpsc::UnboundedSender<JsValue> {
+        self.connect_tx.clone()
+    }
+
+    pub fn respond_to(&self, client: ClientId, response: WorkerResponse) {
+        trace!("Responding to {client:?}");
+        if let Err(e) = self.ports[client.0].send(&response) {
+            error!("Failed to send response to client: {e}");
+        }
+    }
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum TypedMessagePortError {
+    #[error("Could not convert JsValue: {0}")]
+    FailedToConvertValue(serde_wasm_bindgen::Error),
+}
+
+pub struct RequestResponse {
+    port: MessagePortLike,
+    response_channel: Mutex<mpsc::Receiver<Result<WorkerResponse, TypedMessagePortError>>>,
+    _onmessage: Closure<dyn Fn(MessageEvent)>,
+}
+
+impl RequestResponse {
+    pub fn new(object: JsValue) -> Result<Self> {
+        let (tx, rx) = mpsc::channel(1);
+
+        let onmessage = Closure::new(move |ev: MessageEvent| {
+            let response_tx = tx.clone();
+            spawn_local(async move {
+                let message: Result<WorkerResponse, _> =
+                    from_value(ev.data()).map_err(TypedMessagePortError::FailedToConvertValue);
+
+                if let Err(e) = response_tx.send(message).await {
+                    error!("message forwarding channel closed, should not happen: {e}");
+                }
+            })
+        });
+
+        let port = prepare_message_port(object, &onmessage)
+            .context("failed to setup port for RequestResponse")?;
+
+        web_sys::console::log_1(&port);
```

Comment:
in case this is needed and not for debugging, then please attach some message and use tracing